lib/internal/modules/esm/loader.js

```diff
@@ -408,41 +467,94 @@ class ESMLoader {
 
     const namespaces = await PromiseAll(new SafeArrayIterator(jobs));
 
-    return wasArr ?
-      namespaces :
-      namespaces[0];
+    if (!wasArr) return namespaces[0];
+
+    for (let i = 0; i < count; i++) {
+      const namespace = ObjectCreate(null);
+      namespace.url = specifiers[i];
+      namespace.exports = namespaces[i];
+
+      namespaces[i] = namespace;
+    }
+
+    return namespaces;
   }
 
   /**
    * Provide source that is understood by one of Node's translators.
    *
-   * The internals of this WILL change when chaining is implemented,
-   * depending on the resolution/consensus from #36954
-   * @param {string} url The URL/path of the module to be loaded
+   * @param {URL['href']} url The URL/path of the module to be loaded
    * @param {object} context Metadata about the module
-   * @returns {object}
+   * @returns {{ format: ModuleFormat, source: ModuleSource }}
    */
   async load(url, context = {}) {
-    const defaultLoader = this.#loaders[0];
+    const loaders = this.#loaders;
+    let i = loaders.length - 1;
+    let {
+      fn: loader,
+      url: loaderFilePath,
+    } = loaders[i];
+    let chainFinished = i === 0;
+
+    function next(nextUrl, ctx = context) {
+      ({
+        fn: loader,
+        url: loaderFilePath,
+      } = loaders[--i]);
+
+      if (!i) chainFinished = true;
+
+      const hookErrIdentifier = `${loaderFilePath} "load"`;
+
+      if (typeof nextUrl !== 'string') { // non-strings can be coerced to a url string
+        throw new ERR_INVALID_ARG_TYPE(
+          `${hookErrIdentifier} next(url)`,
+          'a url string',
+          nextUrl,
+        );
+      }
 
-    const loader = this.#loaders.length === 1 ?
-      defaultLoader :
-      this.#loaders[1];
-    const loaded = await loader(url, context, defaultLoader);
+      try {
+        new URL(nextUrl);
+      } catch (err) { // eslint-disable-line no-unused-vars
+        throw new ERR_INVALID_ARG_VALUE(
+          `${hookErrIdentifier} next(url)`,
+          nextUrl,
+          'should be a url string',
+        );
+      }
+
+      validateObject(ctx, `${hookErrIdentifier} next(, context)`);
+
+      return loader(nextUrl, ctx, next);
+    }
+
+    const loaded = await loader(
+      url,
+      context,
+      next,
+    );
+
+    const hookErrIdentifier = `${loaderFilePath} load`;
 
     if (typeof loaded !== 'object') {
       throw new ERR_INVALID_RETURN_VALUE(
-        'object',
-        'loader load',
+        'an object',
+        hookErrIdentifier,
         loaded,
       );
     }
```

Comment:
Can we use `validateObject` here?

Reply:
The error thrown by `validateObject()` is different/wrong for this 😞 (I think `INVALID_ARG_TYPE`? Instead of `INVALID_RETURN_TYPE`)

I could catch validateObject's error and re-throw here, but that seemed more trouble than worth. Happy to switch to that if you prefer.

Reply:
I considered the same for validateString() etc, but didn't use them for the same reason (wrong error thrown).

Reply:
Ah right! Maybe we can add a comment explaining that.